### ISSUE TITLE
Passing extra arguments to optimization methods

### ISSUE THREAD

**Issue by FedericoV** [openopt/copt#30]
Most optimization methods do not allow the user to pass extra parameters to the optimization function - requiring the user to instead provide the parameters through a functor or a closure.

Unfortunately, this does not play nicely with the Pytorch JIT, which requires that all input parameters be passed explicitly to the function.  For example:

```
@torch.jit.script
def predict_X(alpha: torch.Tensor,
              B: torch.Tensor,
              X: torch.Tensor,
              allowed_indexes: torch.Tensor,
              t0: int):
    
    t, m = X.shape
    X_pred = torch.zeros((t, m))
    
    for i in range(m):
        Xi_slice = X[:, allowed_indexes[i, :]]
        Bi_slice = B[i, allowed_indexes[i, :]]
        Ai_slice = alpha[allowed_indexes[i, :]]
        
        X_pred[:, i] = Xi_slice @ Bi_slice
        X_pred[t0:, i] +=  Xi_slice[t0:] @ Ai_slice
        
    return X_pred

@torch.jit.script
def calculate_mse(alpha: torch.Tensor,
                  B: torch.Tensor,
                  X: torch.Tensor,
                  allowed_indexes: torch.Tensor,
                  t0: int):
    X_pred = predict_X(alpha, B, X, allowed_indexes, t0)
    res = (X - X_pred)
    mse = torch.mean(res**2)
    return mse


def make_loss_fcn(X: torch.Tensor,
                  allowed_indexes: torch.Tensor,
                  t0: int):
    X = torch.as_tensor(X)
    allowed_indexes = torch.as_tensor(allowed_indexes)
      
    #@torch.jit.script
    def mse_fcn(theta: np.array):
        # We pack all the parameters (alpha and beta) into
        # a single vector of dimension m*(m+1)
        theta = torch.tensor(theta, requires_grad=True)
        
        # solve for m with quadratic: a=1, b=1
        c = theta.shape[0]
        m = int((torch.sqrt(1+4*c)-1) / 2)
        
        # unpack theta from 1d to actual model parameters.
        B = theta[:m*m].reshape((m, m))
        alpha = theta[m*m:]
        
        mse = calculate_mse(alpha, B, X, allowed_indexes, t0)
        mse.backward()
        # return mse, theta.grad
        return mse.detach().clone().numpy(), theta.grad.detach().clone().numpy()

    return mse_fcn

f_grad = make_loss_fcn(torch.as_tensor(X),
                         torch.as_tensor(allowed_indexes),
                         t0)
sol = cp.minimize_proximal_gradient(f_grad, theta, prox=l1_ball.prox, callback=cb)
```

This works: but requires an annoying pattern of creating a closure which creates a loss function which then calls into a jitted function.   One way to fix it would be to allow the user to pass function_args and function_kwargs which are passed through to f_grad automatically.

**Comment by fabianp** [comment]
Should be done in e5c5ccb1fa0911a2035c2e6366c0ecbf28de3eb3 , please let me know if this improves your code
